HoRNDIS.cpp

```diff
@@ -773,66 +723,83 @@ bool HoRNDIS::configureInterface(IONetworkInterface *netif) {
 IOReturn HoRNDIS::getPacketFilters(const OSSymbol *group, UInt32 *filters) const {
 	IOReturn	rtn = kIOReturnSuccess;
 	
-	if (group == gIOEthernetWakeOnLANFilterGroup)
+	if (group == gIOEthernetWakeOnLANFilterGroup) {
 		*filters = 0;
-	else if (group == gIONetworkFilterGroup)
-		*filters = kIOPacketFilterUnicast | kIOPacketFilterBroadcast | kIOPacketFilterMulticast | kIOPacketFilterPromiscuous;
-	else
+	} else if (group == gIONetworkFilterGroup) {
+		// Not yet supporting Multicast: need more code for proper support.
+		// kIOPacketFilterMulticast
+		*filters = kIOPacketFilterUnicast | kIOPacketFilterBroadcast
+			| kIOPacketFilterPromiscuous;
+	} else {
 		rtn = super::getPacketFilters(group, filters);
+	}
 
 	return rtn;
 }
 
 IOReturn HoRNDIS::getMaxPacketSize(UInt32 * maxSize) const {
-	*maxSize = mtu;
+	IOReturn rc = super::getMaxPacketSize(maxSize);
+	if (rc != kIOReturnSuccess) {
+		return rc;
+	}
+	// The max packet size is limited by RNDIS max transfer size:
+	*maxSize = min(*maxSize, maxOutTransferSize - sizeof(rndis_data_hdr));
+	LOG(V_DEBUG, "returning %d", *maxSize);
 	return kIOReturnSuccess;
 }
 
 IOReturn HoRNDIS::selectMedium(const IONetworkMedium *medium) {
+	LOG(V_DEBUG, ">");
 	setSelectedMedium(medium);
 	
 	return kIOReturnSuccess;
 }
 
 IOReturn HoRNDIS::getHardwareAddress(IOEthernetAddress *ea) {
+	LOG(V_DEBUG, ">");
 	UInt32	  i;
 	void *buf;
 	unsigned char *bp;
 	int rlen = -1;
 	int rv;
 	
-	buf = IOMalloc(RNDIS_CMD_BUF_SZ);
-	if (!buf)
+	buf = IOMallocAligned(RNDIS_CMD_BUF_SZ, sizeof(void *));
+	if (!buf) {
 		return kIOReturnNoMemory;
-	
+	}
+
+	// WARNING: Android devices may randomly-generate RNDIS MAC address.
```

Comment:
Yeah, this behavior is super obnoxious.  It leads to quite the proliferation of devices in System Preferences sometimes.

Reply:
That's actually pretty weird. If you call:
attachInterface((IONetworkInterface **)&fNetworkInterface, true),
then things work well. But if you do:
attachInterface((IONetworkInterface **)&fNetworkInterface, false), 
the devices start to proliferate.
... shrug ...